src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -221,15 +273,19 @@ public void OnResponseHeadersComplete(bool endStream)
                 bool signalWaiter;
                 lock (SyncObject)
                 {
-                    if (_state != StreamState.ExpectingHeaders && _state != StreamState.ExpectingTrailingHeaders)
+                    if (_state != StreamState.ExpectingHeaders && _state != StreamState.ExpectingTrailingHeaders && _state != StreamState.ExpectingIgnoredHeaders)
                     {
-                        throw new Http2ProtocolException(Http2ProtocolErrorCode.ProtocolError);
+                        throw new Http2ProtocolException(SR.Format(SR.net_http_http2_protocol_state, "headers", _state));
                     }
 
                     if (_state == StreamState.ExpectingTrailingHeaders || endStream)
                     {
                         _state = StreamState.Complete;
                     }
+                    else if (_state == StreamState.ExpectingIgnoredHeaders)
```

Comment:
It seems like there's an error case we're not detecting here: 1xx response (i.e. ExpectingIgnoredHeaders) with endStream == true. If this happens, it seems like we will get into a bad state. I think we should detect this and throw a protocol exception.

Reply:
I don't think your change fixed the issue. We're checking for endStream above, so we'll never execute the endStream == true case below.

Reply:
BTW, it also seems to me like it's an error if _state == ExpectingTrailingHeaders and endStream is false. We're not catching this today, but it seems like we should.

Reply:
updated.